Keas.Core/Services/IEmailService.cs

```diff
@@ -70,200 +51,145 @@ public async Task SendExpiringMessage(int personId, ExpiringItemsEmailModel mode
-                }
-                _dbContext.KeySerialAssignments.Update(key.KeySerialAssignment);
+                SetNextNotification(assignment);
             }
 
-            foreach (var equipment in expiringItems.Equipment)
+            foreach (var assignment in expiringItems.Equipment.Select(k => k.Assignment))
             {
-                if (equipment.Assignment.NextNotificationDate == null || equipment.Assignment.ExpiresAt > DateTime.UtcNow.AddDays(7))
-                {
-                    equipment.Assignment.NextNotificationDate = equipment.Assignment.ExpiresAt.AddDays(-7);
-                }
-                else if (equipment.Assignment.ExpiresAt > DateTime.UtcNow.AddDays(1))
-                {
-                    equipment.Assignment.NextNotificationDate = equipment.Assignment.ExpiresAt.AddDays(-1);
-                }
-                else
-                {
-                    equipment.Assignment.NextNotificationDate = DateTime.UtcNow.AddDays(1);
-                }
-                _dbContext.EquipmentAssignments.Update(equipment.Assignment);
+                SetNextNotification(assignment);
             }
 
-            foreach (var workstation in expiringItems.Workstations)
+            foreach (var assignment in expiringItems.Workstations.Select(k => k.Assignment))
             {
-                if (workstation.Assignment.NextNotificationDate == null || workstation.Assignment.ExpiresAt > DateTime.UtcNow.AddDays(7))
-                {
-                    workstation.Assignment.NextNotificationDate = workstation.Assignment.ExpiresAt.AddDays(-7);
-                }
-                else if (workstation.Assignment.ExpiresAt > DateTime.UtcNow.AddDays(1))
-                {
-                    workstation.Assignment.NextNotificationDate = workstation.Assignment.ExpiresAt.AddDays(-1);
-                }
-                else
-                {
-                    workstation.Assignment.NextNotificationDate = DateTime.UtcNow.AddDays(1);
-                }
-                _dbContext.WorkstationAssignments.Update(workstation.Assignment);
+                SetNextNotification(assignment);
             }
 
-            foreach (var access in expiringItems.AccessAssignments)
+            foreach (var assignment in expiringItems.AccessAssignments)
             {
-                if (access.NextNotificationDate == null || access.ExpiresAt > DateTime.UtcNow.AddDays(7))
-                {
-                    access.NextNotificationDate = access.ExpiresAt.AddDays(-7);
-                }
-                else if (access.ExpiresAt > DateTime.UtcNow.AddDays(1))
-                {
-                    access.NextNotificationDate = access.ExpiresAt.AddDays(-1);
-                }
-                else
-                {
-                    access.NextNotificationDate = DateTime.UtcNow.AddDays(1);
-                }
-                _dbContext.AccessAssignments.Update(access);

```

Comment:
@jpknoll Was the update statement not needed?

Reply:
No, the objects are already in tracking. Just need the save at the end.

Reply:
@jpknoll Needed, probably because the initial query has .AsNoTracking()

Reply:
Then we should remove the AsNoTracking, because we want to track them :-D

Reply:
@jpknoll I think that caused a lot of other problems. But maybe @jcubbage knows for sure